iptables: ensure all rules are installed consistently

jibi opened the pull request:
See individual commits

jibi committed:
iptables: make IptablesManager thread safe

as its methods can be called concurrently from different subsystems.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: retry installing rules on failure

This commit ensures that in case of any error the InstallRules() and
InstallProxyRules() functions will keep retrying a certain amount of
times before returning an error to the caller.

Since there's going to be an exponential wait time between retries, this
commit moves also the IptablesManager lock from
InstallRules()/InstallProxyRules() to
doInstallRules()/doInstallProxyRules().

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: move custom chains logic into custom_chain.go

This commit moves all the logic related to custom chains into its own
custom_chain.go file. No functional changes are introduced.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: don't ignore errors

This commit ensure that no errors are ignored while installing iptables
rules.

Currently it's possible we miss some errors, for example when renaming
the Cilium ruleset to the "OLD_" prefix, which can in turn cause
failures in installing subsequent rules and thus inconsistencies in the
iptables ruleset managed by Cilium.

This change will allow in a subsequent commit to retry installing the
whole iptables ruleset in case of any failure, to ensure the ruleset is
always in a consistent state.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

christarazi commented on pkg/datapath/iptables/custom_chain.go:
```diff
@@ -94,87 +97,157 @@ var ciliumChains = []customChain{
+			return false, nil
+		}
+
+		return false, fmt.Errorf("unable to list %s chain: %s (%s)", c.name, string(output), err)
```
`%w` for errors

christarazi commented on pkg/proxy/proxy.go:
```diff
@@ -289,9 +289,8 @@ func (p *Proxy) ackProxyPort(name string, pp *ProxyPort) error {
-		if err != nil {
-			return fmt.Errorf("Cannot install proxy rules for %s: %s", name, err)
+		if err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name); err != nil {
+			return fmt.Errorf("Cannot install proxy rules for %s: %w", name, err)
```
Nit: error strings should start lower-cased

christarazi commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -881,6 +883,28 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
+		return nil
+	}
+
+	return nil
```
Shouldn't this return the last error that occurred?

jibi committed:
iptables: refactor addProxyRules

loop over L4 protocols to install rules rather than hardcoding all cases

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: refactor InstallNoTrackRules

No need to pass the ingress parameter to endpointNoTrackRules as we are
always installing rules for both ingress and egress.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: invert conditions to simplify logic

Invert the matching condition in removeCiliumRules, doCopyProxyRule and
addProxyRules to allow continuing to the next rule and reduce branching.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: add debug logs for all commands

Rather than inconsistently logging some of the iptables/ipset
operations, add debug logs for all commands (i.e. all runProg() and
runProgCombinedOutput() calls).

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

christarazi commented on pkg/proxy/proxy.go:
```diff
@@ -289,9 +289,8 @@ func (p *Proxy) ackProxyPort(name string, pp *ProxyPort) error {
-		err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name)
-		if err != nil {
-			return fmt.Errorf("Cannot install proxy rules for %s: %s", name, err)
+		if err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name); err != nil {
```
I think `Install{,Proxy}Rules` should take the daemon `context` so that we can stop the retries if they're in the middle of executing when the daemon is shutting down. 

See calls to `d.Datapath().Loader().Reinitialize()` for reference.

jibi replied:
added a `context` to both `InstallRules` and `InstallProxyRules` :+1: 
The latter was a bit involved as it's called also from `Proxy.ackProxyPort()`, which in turn is called from a few different places

christarazi reviewed (changes requested):
Quick first-glance review. Will take a deeper look when you mark it ready for review.

jibi committed:
iptables: pass context to InstallRules and InstallProxyRules

in order to allow stopping the retries if they're in the middle of
executing when the daemon is shutting down.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: refactor removeRules

No need to loop twice for v4 and v6 rules as the 2 protocols share the
same tables.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: add ruleReferencesDisabledChain and isDisabledChain helpers

which allow to slightly simplify the logic around skipping operations
for the disabled chains.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

christarazi commented on pkg/datapath/iptables/custom_chain.go:
```diff
@@ -94,87 +97,157 @@ var ciliumChains = []customChain{
+	if feedArgs != "" {
+		argsList, err := shellwords.Parse(feedArgs)
+		if err != nil {
+			log.WithError(err).WithField(logfields.Object, feedArgs).Fatal("Unable to parse rule into argument slice")
```
Fatal here generally seems out of place. Why does this failure get treated differently? AFAIK, we usually keep Fatal to the daemon subsystem.

jibi replied:
I do agree that fatal is a bit out of place here, but at the same time I was not sure what the best approach would be as we cannot recover from this error by simply rerunning `doInstallFeeder`, and if we just log it and continue, the iptables ruleset will be most certainly in an inconsistent state (so better fatal out loud and make the error condition clear).

What I can do is just return the error (to remain consistent with the rest of the logic here) and then let the daemon return a fatal error after the 3 attempts

christarazi commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -213,15 +233,8 @@ func (m *IptablesManager) removeCiliumRules(table string, prog iptablesInterface
-		}
-		if skipFeeder {
+		if skip, disabledChain := ruleReferencesDisabledChain(rule); skip {
+			log.WithField("chain", disabledChain).Info("Skipping the removal of feeder chain")
```
Nit: let's add a logfield

jibi committed:
iptables: always try to remove old IPv6 rules

Currently we try to remove old IPv6 rules only if IPv6 support is
enabled, which means we'll leave old IPv6 rules around in case the agent
is restarted after disabling IPv6 support.

What we should do is always try to remove old IPv6 rules, and check only
for support for ip6tables.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
iptables: remove old rules before deleting ipsets

otherwise we may end up trying to delete an ipset which is still
referenced by an old rule, which in turn will raise an error.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
logfields: add Chain constant

used for iptables chains

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -438,148 +455,167 @@ func (m *IptablesManager) installStaticProxyRules() error {
-		log.WithFields(logrus.Fields{
-			"table": table,
-			"prog":  prog.getProg(),
-		}).WithError(err).Warning("Cannot list rules in table. Layer 7 proxy port may get reallocated, which could cause disruption for traffic selected by L7 policy", prog.getProg(), table)
```
Should we keep this information even if we don't emit a warning but return an error instead?

jibi replied:
I don't think the log is relevant anymore as now in case of failure we keep retrying instead of just ignoring the error (and we would also log an error message related to the failure of the list command)

pchaigno replied:
Ok, but if we retry X times and eventually fail for good, what's the error message we will show? Will it highlight the potential disruption of traffic selected by L7 policies? (:point_left: that's the information I fear is now missing for users.)

Is there any downside to passing a new error with that message and the `err` instead of just `err`?

jibi replied:
> Is there any downside to passing a new error with that message and the err instead of just err?

no downside, I just added additional context about L7 policies traffic disruption to the error returned by `copyProxyRule`

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -590,40 +626,44 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
-					"table":          table,
-					"prog":           prog.getProg(),
-					logfields.Object: rule,
-				}).WithError(err).Warn("Unable to copy TPROXY rule, disruption to traffic selected by L7 policy possible")
```
Same here. It seems we'll be loosing some context.

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -590,40 +626,44 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
 			}
+
 			copyRule := append([]string{"-t", table}, args...)
-			log.WithField(logfields.Object, logfields.Repr(copyRule)).Debugf("Copying %s TPROXY rule from %s to %s", prog, oldChain, newChain)
```
Why was this removed?

jibi replied:
After adding the debug logs in `runProg()`/`runProgCombinedOutput()` I thought it would be redundant to keep these logs.

Sure, we'll see an error that only mentions the actual iptables command that failed rather than `Copying [..] TPROXY rule from $OLD_CHAIN to $CHAIN`, but that's probably not an issue as I think we can get already all the context we need just from the plain iptables command that failed (also, if we end up wrapping the error with the additional TPROXY context, why shouldn't we do that with every other rule?)

pchaigno replied:
> After adding the debug logs in `runProg()`/`runProgCombinedOutput()` I thought it would be redundant to keep these logs.

I agree. But when I was reviewing this commit, I didn't know about the last commit which actually has the related changes :wink: :wink: :wink:

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -590,40 +626,44 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
-	output, err := prog.runProgCombinedOutput([]string{"-t", "mangle", "-S"}, true)
+	rules, err := prog.runProgCombinedOutput([]string{"-t", "mangle", "-S"})
 	if err != nil {
-		log.WithError(err).Warnf("Unable to list %s TPROXY rules: %s", prog, string(output))
```
We'll be loosing some context here as well, no?

jibi replied:
Same here: I think we should rely only on the debug logs in `runProg()`/`runProgCombinedOutput()`

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -642,20 +682,20 @@ func (m *IptablesManager) addProxyRules(prog iptablesInterface, proxyPort uint16
 			}
+
 			deleteRule := append([]string{"-t", "mangle"}, args...)
-			log.WithField(logfields.Object, logfields.Repr(deleteRule)).Debugf("Deleting stale %s TPROXY rule from mangle table", prog)
```
Why was this removed?

jibi replied:
(same reason)

pchaigno commented on pkg/datapath/iptables/iptables_test.go:
```diff
@@ -107,6 +107,9 @@ func init() {
 	mockIp4tables.expectations = []expectation{
+		{
+			args: "-t mangle -L CILIUM_PRE_mangle",
+		},
```
Why is this change needed? I expected this commit to have no functional impact on tests :thinking:

jibi replied:
In order to not ignore any meaningful error, when we rename a chain we need to first test if the chain exists (so that we don't return an error in case it doesn't)

pchaigno replied:
Oh, I see. Thanks!

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -850,6 +852,33 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
 
+	backoff := backoff.Exponential{
+		Min:  20 * time.Second,
+		Max:  3 * time.Minute,
```
Doesn't this mean we'll hold the lock above for 3 minutes in the worst case? :grimacing:

jibi replied:
Correct, in the worst case no one else will be able to update the ruleset for the 3 minutes window (and the caller will block for that time). At the same time I'm not sure if making it async is a good strategy, as to my understanding the caller will assume that once this function returns the rules are in place (unless there are errors)

edit: synced on slack: will keep it sync for the first initialization and then make it async for subsequent calls to `InstallRules`/`InstallProxyRules` (but in a separate PR to avoid adding even more changes here)

pchaigno replied:
What if we moved the `m.Lock()` & `m.Unlock()` around `m.doInstallProxyRules()` inside the loop? Would there be any downside to that? It would avoid holding a lock while we're sleeping.

jibi replied:
Good shout :+1: I can't see any downside of that as we are not accessing any `IptablesManager` state in `InstallRules`/`InstallProxyRules` so I moved the locks inside `doInstallRules`/`doInstallProxyRules`

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -668,39 +668,37 @@ func (m *IptablesManager) addProxyRules(prog iptablesInterface, proxyPort uint16
-			}
+		deleteRule := append([]string{"-t", "mangle"}, args...)
+		if err := prog.runProg(deleteRule); err != nil {
+			return err
```
Seems like this change (from `iptables: refactor addProxyRules`) better belongs in the previous commit (`iptables: refactor removeCiliumRules`).

jibi replied:
given that this is about refactoring `addProxyRules` code, I would leave it here

pchaigno replied:
Fine by me, it's a nit really. I'll just note that organizing code changes according to which functions/files they change doesn't do anything to help understand/review.

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -614,23 +614,24 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
+
+		copyRule := append([]string{"-t", table}, args...)
+		if err := prog.runProg(copyRule); err != nil {
+			return err
```
This commit could probably be merged with the other that inverts conditions to simplify code. And instead of the usual `iptables: refactor X`, the title could then be `iptables: invert conditions to simplify logic`.

pchaigno reviewed (changes requested):
A bunch of comments from me, but I don't think there's anything major apart maybe from the long lock holding.

I wonder if it would be best to split this into multiple PRs to ease backports if they are needed:
- First PR with the commits to ensure we error in case of... well errors :grin: 
- Second PR with the refactoring changes, optionally based on the first PR.
- Optionally, we could also move commits `iptables: always try to remove old IPv6 rules` and `iptables: remove old rules before deleting ipsets` to their own PRs if we think they may need to be backported as well.

ldelossa commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -873,77 +805,85 @@ func (m *IptablesManager) endpointNoTrackRules(prog iptablesInterface, cmd strin
+		Max:  3 * time.Minute,
+		Name: "iptables-proxy-rules-installer",
+	}
+
```
nit: Any reason you don't make this a loop variable?

jrajahalme commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -404,6 +405,8 @@ var ciliumChains = []customChain{
 // IptablesManager manages the iptables-related configuration for Cilium.
 type IptablesManager struct {
+	lock.Mutex
+
```
It would be good to document why this lock is needed while `IptablesManager` itself is not modified after `Init()` and why only some of the exported functions take the lock. Apparently this is due to maybe `Install` and `Remove` being called for the same underlying rules concurrently?

jibi replied:
I'll add a comment :+1: 

> Apparently this is due to maybe Install and Remove being called for the same underlying rules concurrently?

correct, we need this look to ensure there are no multiple instances of `InstallRules()`/`InstallProxyRules()` running concurrently as otherwise we may end up with errors or an inconsistent ruleset

ldelossa reviewed (approved):
LGTM, a very dumb nit that you are free to ignore

jrajahalme reviewed (approved):
Generally it would ease backporting if bug fix commits would be separated out from refactoring to a PR of their own. With them interleaved like this it is easy to make a backporting error in one of the refactors if any of the moved code is different between versions.

It might be best to backport this PR separately from the normal backport process to avoid mistakes.
